pkg/apiserver/utils/stats/clickhouse_stats.go

```diff
@@ -0,0 +1,195 @@
+		(intDiv(toUInt32(date_trunc('minute', toDateTime(event_time))), 2) * 2) * 1000 as t,
+		TRUNCATE(avg(ProfileEvent_InsertedRows),0) as RowsPerSecond,
+		formatReadableSize(avg(ProfileEvent_InsertedBytes)) as BytesPerSecond,
+		ROW_NUMBER() OVER(PARTITION BY shardNum() ORDER BY t DESC) rowNumber
+	FROM cluster('{cluster}', system.metric_log)
+	GROUP BY t, shardNum()
+	ORDER BY t DESC, shardNum()
+	) sd
+WHERE sd.rowNumber=1`,
+	stackTraceQuery: `
+SELECT
+	shardNum() as Shard,
+	arrayStringConcat(arrayMap(x -> demangle(addressToSymbol(x)), trace), '\\n') AS trace_function,
+	count()
+FROM cluster('{cluster}', system.stack_trace)
+GROUP BY trace_function, Shard
+ORDER BY count()
+DESC SETTINGS allow_introspection_functions=1`,
+}
+
+type ClickHouseStatQuerierImpl struct {
+	kubeClient        kubernetes.Interface
+	clickhouseConnect *sql.DB
+}
+
+func NewClickHouseStatQuerierImpl(
+	kubeClient kubernetes.Interface,
+) *ClickHouseStatQuerierImpl {
+	c := &ClickHouseStatQuerierImpl{
+		kubeClient: kubeClient,
+	}
+	return c
+}
+
+func (c *ClickHouseStatQuerierImpl) GetDiskInfo(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(diskQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting diskInfo from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) GetTableInfo(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(tableInfoQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting tableInfo from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) GetInsertRate(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(insertRateQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting insertRate from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) GetStackTrace(namespace string, stats *v1alpha1.ClickHouseStats) error {
+	err := c.getDataFromClickHouse(stackTraceQuery, namespace, stats)
+	if err != nil {
+		return fmt.Errorf("error when getting stackTrace from clickhouse: %v", err)
+	}
+	return nil
+}
+
+func (c *ClickHouseStatQuerierImpl) getDataFromClickHouse(query int, namespace string, stats *v1alpha1.ClickHouseStats) error {
+	var err error
+	if c.clickhouseConnect == nil {
+		c.clickhouseConnect, err = util.SetupClickHouseConnection(c.kubeClient, namespace)
+		if err != nil {
+			return err
+		}
+	}
+	result, err := c.clickhouseConnect.Query(queryMap[query])
+	if err != nil {
+		c.clickhouseConnect = nil
+		return fmt.Errorf("failed to get data from clickhouse: %v", err)
+	}
+	defer result.Close()
+	for result.Next() {
+		var err error
+		switch query {
+		case diskQuery:
+			res := v1alpha1.DiskInfo{}
+			err = result.Scan(&res.Shard, &res.Database, &res.Path, &res.FreeSpace, &res.TotalSpace, &res.UsedPercentage)
+			res.UsedPercentage = res.UsedPercentage + " %"
```

Comment:
Is it worth checking err!=nil before this line?

Reply:
Fixed. Thanks.